Created a new component "Story node tile"

pritam2317 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #M1.3
2. This PR does the following: Created a new component "Story node tile"
3. (For bug-fixing PRs only) The original bug occurred because: [Explain what
   the cause of the bug was, and which PR introduced it]

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [X] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [X] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ ] I have written tests for my code.
- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [X] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
![Screenshot from 2024-06-29 03-05-16](https://github.com/oppia/oppia/assets/72196402/1c452e0c-0141-403c-9306-b9c544c8b6d3)


<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->
![Screenshot from 2024-06-29 03-05-33](https://github.com/oppia/oppia/assets/72196402/4b71c2f0-c9b1-40ed-ba04-cc61ba065bbf)

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

pritam2317 committed:
Add feature flag

pritam2317 committed:
create new component storynodetile

pritam2317 committed:
Fix grammar error

pritam2317 committed:
remove unecessary files

pritam2317 committed:
Created a new component Story node tile

pritam2317 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into story_node_tile

Nik-09 reviewed (changes requested):
Hi @pritam2317 please update the PR, since it contains some old changes (related to the feature flag). Thank you

pritam2317 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into story_node_tile

pritam2317 committed:
Push frontend test case

pritam2317 committed:
add copyright and description

Nik-09 commented:
Hi @pritam2317 please assign me once this is ready for review. Thank you

pritam2317 commented:
> Hi @pritam2317 please assign me once this is ready for review. Thank you

Hi @Nik-09, now I've assigned you.

Nik-09 reviewed (changes requested):
Hi @pritam2317 There are a few UI feedbacks:

![image](https://github.com/oppia/oppia/assets/57531197/0f4fd841-4e74-4765-b04d-397c2825d4c6)

As per the above Figma mock, 
1. Padding is missing around texts.
2. The start button should be a little larger in size with a white background.
3. The mobile view of the card does not exactly match the figma mocks; please update.

pritam2317 commented:
> Hi @pritam2317 There are a few UI feedbacks:
> 
> ![image](https://private-user-images.githubusercontent.com/57531197/342693789-0f4fd841-4e74-4765-b04d-397c2825d4c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzOTUyNTAsIm5iZiI6MTcxOTM5NDk1MCwicGF0aCI6Ii81NzUzMTE5Ny8zNDI2OTM3ODktMGY0ZmQ4NDEtNGU3NC00NzY1LWIwNGQtMzk3YzI4MjVkNGM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDA5NDIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1YWYzYmViYWIxMmYyMTEyYjFhYmFkZjg2MmU1MTBmNmNlOTllNzIwYjI3MmM4OTc4NTA4NTMxYzE1NDg3NzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B6xT93mMzfHtJ0Rh59vBhnBGPbmlzQchOnP9QIx6Sng)
> 
> As per the above Figma mock, 
> 
> 1. Padding is missing around texts.
> 2. The start button should be a little larger in size with a white background.
> 3. The mobile view of the card does not exactly match the figma mocks; please update.

But according to mocks the background color of **start** and **resume** button is green.

pritam2317 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into story_node_tile

pritam2317 committed:
fixes css

pritam2317 committed:
fixes lint check

pritam2317 commented:
Hi @Nik-09  I've updated the video proof. Please take a look at it, and then I will update the code here.

seanlip commented:
Hi @pritam2317, the video in the description is still from 2 weeks ago. 

I would also recommend uploading the video and the code in the same batch to minimize delays in reviewing. It should be possible for you to look at your output and the Figma mocks to see whether your UI matches the mocks.

pritam2317 committed:
Fixes css and typescript checks

pritam2317 committed:
Fixes linter checks

pritam2317 committed:
fixing typescript issue

pritam2317 committed:
fixing typescript

pritam2317 commented:
So sorry @seanlip it was a mistake from my end, but now I've added the screenshot for both desktop and mobile, PTAL

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.spec.ts:
```diff
@@ -0,0 +1,522 @@
+    componentInstance.checkpointCount = 3;
+    componentInstance.totalCheckpointLoaded = true;
+    componentInstance.mostRecentlyReachedCheckpoint = 'Mid';
+  }));
```
add newline

pritam2317 replied:
Done.

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.spec.ts:
```diff
@@ -0,0 +1,522 @@
+    }
+  );
+
+  it('should get the completed progress bard width', () => {
```
```suggestion
  it('should get the completed progress bar width', () => {
```

pritam2317 replied:
Now it's fixed.

Nik-09 reviewed (changes requested):
Thank you, @pritam2317, for the PR. I added a few comments.
Also, can you please update the screenshots in which there should be some images?

pritam2317 committed:
fix typscript error

pritam2317 committed:
fix grammatical error in test file

pritam2317 commented:
> Thank you, @pritam2317, for the PR. I added a few comments. Also, can you please update the screenshots in which there should be some images?

Hi @Nik-09, I've added the screenshots and pushed the code.PTAL

Nik-09 commented:
> > Thank you, @pritam2317, for the PR. I added a few comments. Also, can you please update the screenshots in which there should be some images?
> 
> Hi @Nik-09, I've added the screenshots and pushed the code.PTAL

Hi @pritam2317 the screenshots do not contain images. Can you please update?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.html:
```diff
@@ -0,0 +1,213 @@
+        <div class="completed-progress-bar" [style.width.%]="getCompletedProgressBarWidth()"></div>
+      </div>
+      <div *ngFor="let checkpointNode of checkpointStatusArray; index as idx"
+                [ngClass]="{'completed-checkpoint-node': checkpointStatusArray[idx] === 'completed', 'in-progress-checkpoint-node': checkpointStatusArray[idx] === 'in-progress', 'incomplete-checkpoint-node': checkpointStatusArray[idx] === 'incomplete'}">
```
Why are we doing ```checkpointStatusArray[idx]```, can we directly use checkpointNode inside for loop?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.html:
```diff
@@ -0,0 +1,213 @@
+      </div>
+    </div>
+    <button class="node-button" *ngIf="resumeButtonActivate">resume</button>
+    <button class="node-button" *ngIf="!resumeButtonActivate" (click)="clickstart()">start</button>
```
Please rename the method, it should describe the action a bit. Somewhere around ```StartLesson()```?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.spec.ts:
```diff
@@ -0,0 +1,523 @@
+
+    expect(componentInstance.completedcheckpoints).toBe(1);
+    expect(componentInstance.nodeDataLoaded).toEqual(true);
+    spyOn(componentInstance, 'getCheckpointCount').and.callThrough();
```
Why these Spy's are required at the end of the test case?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.spec.ts:
```diff
@@ -0,0 +1,523 @@
+    componentInstance.ngOnInit();
+    expect(componentInstance.checkpointCount).toEqual(3);
+    expect(componentInstance.totalCheckpointLoaded).toEqual(true);
+    spyOn(componentInstance, 'getCompletedProgressStatus').and.callThrough();
```
Ditto

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.spec.ts:
```diff
@@ -0,0 +1,523 @@
+    componentInstance.completedcheckpoints = 1;
+    expect(componentInstance.getCompletedProgressBarWidth()).toEqual(25);
+    componentInstance.completedcheckpoints = 2;
+    expect(componentInstance.getCompletedProgressBarWidth()).toEqual(75);
```
Help me to understand the math here, if two checkpoints are completed then how come the progress is 75?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+  @Input() storyId!: string;
+  @Input() topicUrlFragment!: string;
+  @Input() classroomUrlFragment!: string;
+  @Input() storyUrlFragment!: string;
```
Add a new line below, to segregate Input variables and defined variables.

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+  totalCheckpointLoaded = false;
+  checkpointStatusArray!: string[];
+  resumeButtonActivate = false;
+  constructor(
```
Add a new line above

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+      .then((response: FetchExplorationBackendResponse) => {
+        this.expStates = response.exploration.states;
+        let count = 0;
+        for (let [, value] of Object.entries(this.expStates)) {
```
If we only need values, then can we use Object.values, instead of Object.entries?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+      });
+  }
+
+  getCompletedProgressStatus(): void {
```
The function name is getCompletedProgressStatus but the function does not return anything, can we rename this?

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+    return iconParametersArray;
+  }
+
+  clickstart(): void {
```
Please rename this, as suggested above.

Nik-09 commented on core/templates/pages/topic-viewer-page/story-node-tile/story-node-tile.component.ts:
```diff
@@ -0,0 +1,243 @@
+    window.open(startLink, '_blank');
+  }
+
+  getStartLink(): string {
```
I don't understand ```startLink```, can we rename this to getLessonUrl()?

Nik-09 reviewed (changes requested):
Added a few more comments.
